existing mirror reconfig should be prevented

sunilangadi2 opened the pull request:
Signed-off-by: Sunil Angadi <sangadi@redhat.com>

# Description
Fixes RHCEPHQE-6712
Test case to cover negative scenario 
Attempt creating mirror in same cluster. This should be prevented

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

rkhambad commented:
> 

Please always add a polarion link while automating test case , it will be easy to check the test case and expected behavior

sunilangadi2 commented:
> > 
> 
> Please always add a polarion link while automating test case , it will be easy to check the test case and expected behavior

Sorry done added link.

sunilangadi2 commented on suites/pacific/rbd/tier-1_rbd_mirror.yaml:
```diff
@@ -166,7 +166,7 @@ tests:
             io-total: 200M
             resize_to: 5G
-      polarion-id: CEPH-83573332
+      polarion-id: CEPH-83573332, CEPH-9511
```
@rkhambad Is it ok if i add those steps in the existing polarion id and make 9511 as Inactive?

sunilangadi2 replied:
@Manohar-Murthy @VasishtaShastry To avoid multiple polarion ID, can we make `CEPH-9511` as Inactive by adding this test case step in `CEPH-83573332`?
Please let me know your thoughts ?
creating a separate test module for this negative scenario is redundant, in this existing mirror setup 
we can cross check this negative scenario

VasishtaShastry replied:
@sunilangadi2 
IMO we can avoid above mentioned TC to be merged
we should block the build if CEPH-83573332 fails but not if CEPH-9511
I see convinience as the only reason being mentioned here to merge, which I'm not able to agree with

sunilangadi2 replied:
@VasishtaShastry Agree with your point, but when test case fails or blocking build once again we will cross check manually Is it at that time can we decide is it?
My only concern is, to cover this negative scenario, we just have repeat some of the existing steps
I though of avoiding
Please let me know your thoughts once again?
i am ok to go with new module.

VasishtaShastry replied:
In general,
-> but when test case fails or blocking build once again we will cross check manually Is it at that time can we decide is it?
cross checking is done on rotation basis, new person would be checking them periodically, i don't see any validation to create this dependency between monitor and SMEs&authors each time.

-> I  think it should be okay if we merge similar/overlapping TCs of same tier

sunilangadi2 replied:
ok sure let me cross check then.

Manohar-Murthy replied:
Please have two seperate tests as scenarios are different. Develop another module for negative scenario.

sunilangadi2 replied:
Done added new module itself.

sunilangadi2 commented on tests/rbd_mirror/test_rbd_mirror.py:
```diff
@@ -54,7 +54,14 @@ def run(**kw):
-        mirror1.clean_up(peercluster=mirror2, pools=[poolname])
+
+        # Attempt creating mirror in same cluster this should be prevented
+        out = mirror1.enable_mirroring(mirror_level="pool", specs=poolname, mode="pool")
```
ok sure i can add this but as test case didn't mention, hence done for only pool based.

sunilangadi2 replied:
Done covered image mode as well

VasishtaShastry commented on tests/rbd_mirror/test_rbd_mirror.py:
```diff
@@ -54,7 +54,14 @@ def run(**kw):
         mirror1.check_data(peercluster=mirror2, imagespec=imagespec)
-        mirror1.clean_up(peercluster=mirror2, pools=[poolname])
+
+        # Attempt creating mirror in same cluster this should be prevented
```
test_rbd_mirror.py is being used as gateway for rbd-mirorr suites
I think it doesn't make sense to add a negative scenario atleast in this module.

sunilangadi2 replied:
ok got your point, what is the suggestion to avoid multiple polarion ID?
Is that new module itself?

VasishtaShastry replied:
For now it appears like yes
if you can comeup with any other solution, that would be interesting

sunilangadi2 commented:
Test Result:
[Passed] http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-WYBH5B
Please review it.

rkhambad commented on tests/rbd_mirror/test_rbd_mirror_reconfig.py:
```diff
@@ -0,0 +1,79 @@
+            rbdmirror.RbdMirror(cluster, config)
+            for cluster in kw.get("ceph_cluster_dict").values()
+        ]
+        poolname = mirror1.random_string() + "_tier_1_rbd_mirror_pool"
```
we can optimize the code using list and loop over here for mode 
Like modes = ["pool", "image"] 
and log msg we can modify accordingly.

sunilangadi2 replied:
Done

manasagowri commented on tests/rbd_mirror/test_rbd_mirror_reconfig.py:
```diff
@@ -0,0 +1,79 @@
+    try:
+        log.info("Starting RBD mirroring test case")
+        config = kw.get("config")
+        mirror1, mirror2 = [
```
Could you please use rbd_mirror_config method so that we can cover ecpool implementation part as well while automating the test case itself.

sunilangadi2 commented:
Test results for updated code:
Passed: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-7F7VMQ

VasishtaShastry reviewed (changes requested):
Really sorry for the belated review.

sunilangadi2 commented on tests/rbd_mirror/test_rbd_mirror_reconfig.py:
```diff
@@ -0,0 +1,101 @@
+        pool = config[pool_type]["pool"]
+        mode = config[pool_type]["mode"]
+        # Attempt creating mirror in same cluster this should be prevented
+        out = mirror1.enable_mirroring(mirror_level="pool", specs=pool, mode=mode)
```
As per the test case bootstrap peers not needed,
It's just about enabling mirroring again to existing mirror setup

VasishtaShastry replied:
the phrases 
1) Step  "Attempt creating mirror"
2) Expected results - "Either the CLI should not provide any option to specify same cluster source and target or it detects the condition where source and target cluster for the mirror same and prints an error message."

Clearly deserves verification of something more than just attempting enabling mirroring on the pool.
Enabling Mirroring doesn't even care about source and destination, it prevents with same error message even if you are trying to configure to different cluster where as title clearly says "Attempt creating mirror in same cluster."

Please let me know your thoughts.

sunilangadi2 replied:
Done Please review

manasagowri commented on tests/rbd_mirror/test_rbd_mirror_reconfig.py:
```diff
@@ -0,0 +1,101 @@
+        log.error(
+            f"{kw.get('ceph_cluster_dict').values} has less or more clusters Than Expected(2 clusters expected)"
+        )
+        log.exception(ve)
```
One small thing. Please add a return 1 here.

sunilangadi2 replied:
Done please review

VasishtaShastry commented on tests/rbd_mirror/test_rbd_mirror_reconfig.py:
```diff
@@ -0,0 +1,113 @@
+            log.error("importing bootstrap token with same cluster not prevented")
+            return 1
+
+        out = mirror1.enable_mirroring(mirror_level="pool", specs=pool, mode=mode)
+        if "mirroring is already configured" in out:
+            log.info(f"creating mirror with same cluster got prevented for {mode} mode")
+        else:
+            log.error(
+                f"creating mirror with same cluster not prevented for {mode} based"
+            )
+            return 1
```
I don't see much value addition here from the perspective of testing prevention of configuring mirroring with same cluster.

pool enable - this command would fail irrespective of same cluster or different cluster.

sunilangadi2 replied:
agree but it would be good to cover that scenario as well right as coding is already done for it 
there may be chance that enabling mirroring to already enabling mirror command might pass if there is a bug
hence to cover that scenario as well,
Kept the remaining code as it is and covered the scenario for the mirroring bootstrap to fail.

VasishtaShastry replied:
Discussed in automation peer review meeting and agreed to keep the code with modifications to the polarion test case.

Please modify the TC in polarion and request for review.

sunilangadi2 replied:
Updated polarion TC please review it.

sunilangadi2 commented:
Please review it addresed all comments:
Test results:
Passed: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-LLEL05

rkhambad commented on tests/rbd_mirror/test_rbd_mirror_reconfig.py:
```diff
@@ -0,0 +1,113 @@
+        primary_cluster_name = primary_cluster.name
+        mirror1.bootstrap_peers(poolname=pool, cluster_name=primary_cluster_name)
+        # Attempt creating mirror in same cluster this should be prevented
+        out, err = mirror1.import_bootstrap(
```
[**optional**] can we log the actual output for import_bootstrap command, to cross verify the message we are logging.

sunilangadi2 replied:
IMO we don't need to log output as it's negative scenario command is failing, it's in `err` output been cross verified in line 32 Hence no need of output log.

rkhambad reviewed (commented):
LGTM

sunilkumarn417 commented on tests/rbd_mirror/rbd_mirror_utils.py:
```diff
@@ -57,7 +57,7 @@ def exec_cmd(self, **kw):
                 cmd = cmd + self.ceph_args
 
-            out = node.exec_command(
+            out, err = node.exec_command(
```
return values could be only one if long_running is passed.

sunilangadi2 replied:
Done

sunilkumarn417 commented on tests/rbd_mirror/rbd_mirror_utils.py:
```diff
@@ -70,6 +70,8 @@ def exec_cmd(self, **kw):
                     return out[0]
                 return out
 
+            if kw.get("all", False):
```
changes needed correspondingly w.r.t to above comment

sunilangadi2 replied:
Done

preethisn commented:
LGTM

rkhambad reviewed (approved):
changes looks good

sunilkumarn417 commented:
- Please provide complete test workflow and automation test results post correction.

sunilangadi2 commented:
Updated Test work flow in test module also updated polarion steps,
Test Results:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-TVDVHE/

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_reconfig.py:
```diff
@@ -0,0 +1,75 @@
+        config = kw.get("config")
+        mirror1, mirror2 = [
+            rbdmirror.RbdMirror(cluster, config)
+            for cluster in kw.get("ceph_cluster_dict").values()
```
please dont use `get`.

sunilangadi2 replied:
Done.

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_reconfig.py:
```diff
@@ -0,0 +1,75 @@
+            return 1
+        return 0
+
+    except ValueError as ve:
```
please take out this exception block

sunilangadi2 replied:
Done.

sunilangadi2 commented:
Test results after addressing above comments:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-0L0B6H

sunilangadi2 commented:
@sunilkumarn417 @manasagowri Repushed the code for tox failure, Please review it.

sunilkumarn417 reviewed (commented):
@sunilangadi2 please rerun the test case with debug Log level and please update the polarion test case.

sunilangadi2 commented:
> @sunilangadi2 please rerun the test case with debug Log level and please update the polarion test case.

Test result with debug log level
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-RRDS08

updated polarion steps as well along with config.
Please review it.

rkhambad reviewed (approved):
changes look good.
